staging/src/k8s.io/apiextensions-apiserver/pkg/storageversion/manager.go

```diff
@@ -0,0 +1,256 @@
+	q := m.getOrCreateUpdateQueue(crd.UID)
+	// When the channel buffer is full, writing to the channel becomes
+	// blocking. Here we give up updating storage version for this CRD and
+	// print a log, so that we can return immediately and not block the
+	// caller.
+	if len(q) == updateQueueBufferSize {
+		// TODO(roycaihw): use warning instead of info when StorageVersionAPI
+		// graduates to beta/GA
+		klog.V(2).Infof("Skipping the storage version update for CRD with UID %v due to the queue being full (queue size: %v).",
+			crd.UID, updateQueueBufferSize)
+		if processedCh != nil {
+			close(processedCh)
+		}
+		return
+	}
+	q <- &update{
+		crd:         crd,
+		waitCh:      waitCh,
+		processedCh: processedCh,
+		abortedCh:   abortedCh,
+	}
+}
+
+// getOrCreateUpdateQueue returns the channel for the given UID, or create a new
+// one and a new goroutine if necessary. The goroutine keeps processing updates
+// until the channel is closed.
+func (m *manager) getOrCreateUpdateQueue(uid types.UID) chan<- *update {
+	if queue, ok := m.updateQueues[uid]; ok {
+		return queue.q
+	}
+
+	queue := make(chan *update, updateQueueBufferSize)
+	ctx, cancel := context.WithCancel(context.TODO())
+	m.updateQueues[uid] = &updateQueue{
+		q:      queue,
+		cancel: cancel,
+	}
+	go func() {
+		for update := range queue {
+			select {
+			case <-ctx.Done():
+				// The queue was cancelled. Abort the update.
+				if update.abortedCh != nil {
+					close(update.abortedCh)
+				}
+				continue
+			default:
+			}
+
+			if update.waitCh != nil {
+				<-update.waitCh
+			}
+			if err := m.updateStorageVersionFor(ctx, update.crd); err != nil {
+				utilruntime.HandleError(err)
+			}
+			if update.processedCh != nil {
+				select {
+				case <-ctx.Done():
+					// The queue was cancelled. Potentially we didn't finish this
+					// storage version update.
+					close(update.abortedCh)
+				default:
+					close(update.processedCh)
+				}
+			}
+		}
+	}()
+	return queue
+}
+
+// TeardownFor closes the channel for the given UID. It ensures that we don't
+// leak goroutines.
+func (m *manager) TeardownFor(uid types.UID) {
+	if queue, ok := m.updateQueues[uid]; ok {
+		// Cancel all the pending updates, so that if the CRD is
+		// re-created, the old updates won't race with the new updates.
+		// We can safely discard the old storage version updates
+		// because:
+		//   1. if the CRD is deleted forever, all CRs will be GC'ed and
+		//      the storage version doesn't matter.
+		//   2. if the CRD gets deleted and re-created in a short period
+		//      and the old CRs remain, the new CRD will update new
+		//      storage version.
+		klog.V(4).Infof("Cancelling the storage version update queue for CRD with UID %v.",
+			uid)
+		queue.cancel()
+		close(queue.q)
+		delete(m.updateQueues, uid)
+	}
+}
+
+func (m *manager) updateStorageVersionFor(ctx context.Context, crd *apiextensionsv1.CustomResourceDefinition) error {
```

Comment:
we have 3 kinds of "UpdateStorageVersionFor" funcs. Maybe rename them to be more clear?